### PR TITLE
Add deviceIdProvider closure for custom device ID generation

### PR DESCRIPTION
**Pull request by jaredmixpanel**
## Summary

Adds a `deviceIdProvider: (() -> String?)?` property to `MixpanelOptions` that allows customers to supply their own device ID generation logic.

**Key behaviors:**
- Provider called on init (if no persisted identity) and after `reset()`/`optOutTracking()`
- Return same value = persistent device ID (never resets)
- Return different value = ephemeral device ID (resets each time)
- Return `nil` = fall back to SDK default (graceful error handling)
- SDK logs error when provider would replace existing persisted `anonymousId`
- Empty string from provider falls back to default UUID/IDFV behavior

This gives customers full control over device ID persistence and reset semantics without needing separate parameters on `reset()` and `optOutTracking()`.

## Motivation

Customer request for persistent device IDs that survive `reset()` calls. Rather than adding boolean flags or additional parameters, this closure-based approach gives customers complete flexibility over:
- Where device IDs are stored (Keychain, server, etc.)
- Whether device IDs reset or persist
- Custom device ID formats
- Graceful fallback when device ID source is unavailable

## Changes

- `MixpanelOptions.swift`: Added `deviceIdProvider` property with `(() -> String?)?` type and comprehensive documentation
- `MixpanelInstance.swift`: Modified `defaultDeviceId()` to check provider and handle nil/empty; added warning logic in `unarchive()`
- `MixpanelDeviceIdProviderTests.swift`: 15 unit tests covering all behaviors including nil return
- `docs/device-id-provider.md`: Comprehensive usage documentation
- `MixpanelDemo/AppDelegate.swift`: QA test providers (persistent, ephemeral, failing)

## API

```swift
let options = MixpanelOptions(
    token: "YOUR_TOKEN",
    deviceIdProvider: {
        // Option 1: Persistent - return same value each time
        return MyKeychainHelper.getOrCreatePersistentId()
        
        // Option 2: Ephemeral - return different value each time
        // return UUID().uuidString
        
        // Option 3: Fallback - return nil on failure
        // guard let id = fetchFromServer() else { return nil }
        // return id
    }
)
```

## Test plan

- [x] Run `MixpanelDeviceIdProviderTests` - 15/15 tests pass
- [x] Run full iOS test suite - all tests pass
- [x] Test nil return fallback behavior
- [x] Manual testing with MixpanelDemo app

## GitHub Copilot Summary

This pull request introduces a customizable device ID provider to the Mixpanel Swift SDK, allowing developers to control how device IDs are generated and managed. This enables advanced identity management strategies, such as persistent device IDs across app reinstalls or ephemeral IDs that change on reset. The changes include SDK API updates, documentation, and demo app enhancements for testing different device ID behaviors.

**Device ID Provider Feature**

* Added a new `deviceIdProvider` closure property to `MixpanelOptions`, allowing developers to specify custom logic for device ID generation. The closure is called during initialization, `reset()`, and `optOutTracking()`, and its behavior is documented in detail. [[1]](diffhunk://#diff-b305cdb279fdbeb711fa36c9b9bc9253025444fc674fa76d0437c99b78b08dc4R23-R75) [[2]](diffhunk://#diff-b305cdb279fdbeb711fa36c9b9bc9253025444fc674fa76d0437c99b78b08dc4L35-R89) [[3]](diffhunk://#diff-b305cdb279fdbeb711fa36c9b9bc9253025444fc674fa76d0437c99b78b08dc4R103) [[4]](diffhunk://#diff-efb0304d15600701ffd641c45103d3bb51a4416d48961469f73f631199ec6ff5R1-R201)
* Updated `MixpanelInstance` logic to use the custom provider when available, with fallback to default behavior if the provider returns `nil` or an empty string. Added logging to warn about identity discontinuity if the provider returns a value different from the persisted anonymous ID. [[1]](diffhunk://#diff-764368484691977ff02e15390a9fac71fa56690bb43a46b1d99ab3e5d6dd50c5R631-R643) [[2]](diffhunk://#diff-764368484691977ff02e15390a9fac71fa56690bb43a46b1d99ab3e5d6dd50c5R1045-R1060)

**Demo App Enhancements**

* Added multiple device ID provider options (`persistent`, `ephemeral`, `failing`, and default) to `AppDelegate.swift` for easy QA and testing of provider behaviors. Console output clarifies which provider is in use and the resulting device IDs.
* Improved the `reset()` flow in `UtilityViewController.swift` to show before/after device IDs in the console and indicate whether the ID changed, aiding QA of provider implementations.

**Documentation**

* Added a comprehensive `device-id-provider.md` guide explaining how to use the new feature, best practices, edge cases, and integration with Mixpanel's identity management.

**Project Structure**

* Added `MixpanelDeviceIdProviderTests.swift` to the test target in the Xcode project for unit testing custom provider logic. [[1]](diffhunk://#diff-118db5f9f9e95f2fe22afb05294f6ae087a8630369ed6fb4088700ff327d80c9R11) [[2]](diffhunk://#diff-118db5f9f9e95f2fe22afb05294f6ae087a8630369ed6fb4088700ff327d80c9R255) [[3]](diffhunk://#diff-118db5f9f9e95f2fe22afb05294f6ae087a8630369ed6fb4088700ff327d80c9R597) [[4]](diffhunk://#diff-118db5f9f9e95f2fe22afb05294f6ae087a8630369ed6fb4088700ff327d80c9R1124)

**Minor Fixes**

* Corrected a typo in the documentation for the `optInTracking()` method parameter (`distintId` → `distinctId`).

### CODE REVIEW

**Comment by jaredmixpanel** [comment]
### Code review

Found 1 issue:

1. Missing thread-safety warning in deviceIdProvider documentation (CLAUDE.md says "Thread Safety is Mandatory")

The `deviceIdProvider` closure is called while holding internal `readWriteLock.write` locks (in `reset()`, `optOutTracking()`, and `unarchive()`), but the documentation doesn't warn users about this constraint. The examples suggest potentially blocking operations like Keychain access (`MyKeychainHelper.getOrCreatePersistentId()`) and network calls (`fetchDeviceIdFromServer()`). If users implement blocking providers, this could cause deadlocks similar to the issue fixed in commit 37da555.

Consider adding a note like: "This closure is called synchronously on internal queues. Avoid blocking operations (Keychain, network, UserDefaults) as they may cause deadlocks."

https://github.com/mixpanel/mixpanel-swift/blob/6a642a8437b345a4c8056f9d360f06d13d18ef6f/Sources/MixpanelOptions.swift#L42-L72

---

Generated with [Claude Code](https://claude.ai/code)

<sub>- If this code review was useful, please react with a thumbs up. Otherwise, react with a thumbs down.</sub>

File: Sources/MixpanelInstance.swift [hunk 1]
```diff
@@ -628,6 +628,19 @@ open class MixpanelInstance: CustomDebugStringConvertible, FlushDelegate, AEDele
   }
 
   func defaultDeviceId() -> String {
+    // Check if a custom device ID provider is set
+    if let provider = options.deviceIdProvider {
```
Comment (ketanmixpanel) [inline Sources/MixpanelInstance.swift]: @jaredmixpanel  Do you know if there is device_id length limitation? if yes, then I think its good idea to add length check here.
